File: app/src/main/java/com/romellfudi/ussd/sample/MainFragment.java
package com.romellfudi.ussd.sample;

import android.Manifest.permission;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.romellfudi.permission.PermissionService;
import com.romellfudi.ussd.R;
import com.romellfudi.ussdlibrary.OverlayShowingService;
import com.romellfudi.ussdlibrary.SplashLoadingService;
import com.romellfudi.ussdlibrary.USSDApi;
import com.romellfudi.ussdlibrary.USSDController;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;

/**
 * Use Case for Test Windows
 *
 * @author Romell Domínguez
 * @version 1.1.b 27/09/2018
 * @since 1.0.a
 */
public class MainFragment extends Fragment {

    private TextView result;
    private EditText phone;
    private Button btn1, btn2, btn3, btn4;
    private HashMap<String, HashSet<String>> map;
    private USSDApi ussdApi;
    private MainActivity menuActivity;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        map = new HashMap<>();
        map.put("KEY_LOGIN", new HashSet<>(Arrays.asList("espere", "waiting", "loading", "esperando")));
        map.put("KEY_ERROR", new HashSet<>(Arrays.asList("problema", "problem", "error", "null")));
        ussdApi = USSDController.getInstance(getActivity());
        menuActivity = (MainActivity) getActivity();
        new PermissionService(getActivity()).request(
                new String[]{permission.CALL_PHONE, permission.READ_PHONE_STATE},
                callback);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.content_op1, container, false);
        result = (TextView) view.findViewById(R.id.result);
        phone = (EditText) view.findViewById(R.id.phone);
        btn1 = (Button) view.findViewById(R.id.btn1);
        btn2 = (Button) view.findViewById(R.id.btn2);
        btn3 = (Button) view.findViewById(R.id.btn3);
        btn4 = (Button) view.findViewById(R.id.btn4);
        setHasOptionsMenu(false);

        btn1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String phoneNumber = phone.getText().toString().trim();
                ussdApi = USSDController.getInstance(getActivity());
                result.setText("");
                ussdApi.callUSSDInvoke(phoneNumber, map, new USSDController.CallbackInvoke() {
                    @Override
                    public void responseInvoke(String message) {
                        Log.d("APP", message);
                        result.append("\n-\n" + message);
                        // first option list - select option 1
                        ussdApi.send("1", new USSDController.CallbackMessage() {
                            @Override
                            public void responseMessage(String message) {
                                Log.d("APP", message);
                                result.append("\n-\n" + message);
                                // second option list - select option 1
                                ussdApi.send("1", new USSDController.CallbackMessage() {
                                    @Override
                                    public void responseMessage(String message) {
                                        Log.d("APP", message);
                                        result.append("\n-\n" + message);
                                    }
                                });
                            }
                        });
                    }

                    @Override
                    public void over(String message) {
                        Log.d("APP", message);
                        result.append("\n-\n" + message);
                    }
                });
            }
        });

        btn2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (USSDController.verifyOverLay(getActivity())) {
                    final Intent svc = new Intent(getActivity(), OverlayShowingService.class);
                    svc.putExtra(OverlayShowingService.EXTRA, "PROCESANDO");
                    getActivity().startService(svc);
                    Log.d("APP", "START OVERLAY DIALOG");
                    String phoneNumber = phone.getText().toString().trim();
                    ussdApi = USSDController.getInstance(getActivity());
                    result.setText("");
                    ussdApi.callUSSDOverlayInvoke(phoneNumber, map, new USSDController.CallbackInvoke() {
                        @Override
                        public void responseInvoke(String message) {
                            Log.d("APP", message);
                            result.append("\n-\n" + message);
                            // first option list - select option 1
                            ussdApi.send("1", new USSDController.CallbackMessage() {
                                @Override
                                public void responseMessage(String message) {
                                    Log.d("APP", message);
                                    result.append("\n-\n" + message);
                                    // second option list - select option 1
                                    ussdApi.send("1", new USSDController.CallbackMessage() {
                                        @Override
                                        public void responseMessage(String message) {
                                            Log.d("APP", message);
                                            result.append("\n-\n" + message);
                                            getActivity().stopService(svc);
                                            Log.d("APP", "STOP OVERLAY DIALOG");
                                            Log.d("APP", "successful");
                                        }
                                    });
                                }
                            });
                        }

                        @Override
                        public void over(String message) {
                            Log.d("APP", message);
                            result.append("\n-\n" + message);
                            getActivity().stopService(svc);
                            Log.d("APP", "STOP OVERLAY DIALOG");
                        }
                    });
                }
            }
        });

        btn4.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (USSDController.verifyOverLay(getActivity())) {
                    final Intent svc = new Intent(getActivity(), SplashLoadingService.class);
                    getActivity().startService(svc);
                    Log.d("APP", "START SPLASH DIALOG");
                    String phoneNumber = phone.getText().toString().trim();
                    result.setText("");
                    ussdApi.callUSSDOverlayInvoke(phoneNumber, map, new USSDController.CallbackInvoke() {
                        @Override
                        public void responseInvoke(String message) {
                            Log.d("APP", message);
                            result.append("\n-\n" + message);
                            // first option list - select option 1
                            ussdApi.send("1", new USSDController.CallbackMessage() {
                                @Override
                                public void responseMessage(String message) {
                                    Log.d("APP", message);
                                    result.append("\n-\n" + message);
                                    // second option list - select option 1
                                    ussdApi.send("1", new USSDController.CallbackMessage() {
                                        @Override
                                        public void responseMessage(String message) {
                                            Log.d("APP", message);
                                            result.append("\n-\n" + message);
                                            getActivity().stopService(svc);
                                            Log.d("APP", "STOP SPLASH DIALOG");
                                            Log.d("APP", "successful");
                                        }
                                    });
                                }
                            });
                        }

                        @Override
                        public void over(String message) {
                            Log.d("APP", message);
                            result.append("\n-\n" + message);
                            getActivity().stopService(svc);
                            Log.d("APP", "STOP OVERLAY DIALOG");
                        }
                    });
                }
            }
        });

        btn3.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                USSDController.verifyAccesibilityAccess(getActivity());
            }
        });

        return view;
    }

    private PermissionService.Callback callback = new PermissionService.Callback() {
        @Override
        public void onRefuse(ArrayList<String> RefusePermissions) {
            Toast.makeText(getContext(),
                    getString(R.string.refuse_permissions),
                    Toast.LENGTH_SHORT).show();
            getActivity().finish();
        }

        @Override
        public void onFinally() {
            // pass
        }
    };

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        callback.handler(permissions, grantResults);
    }
}

File: app-kotlin/src/main/java/com/romellfudi/ussd/sample/MainFragment.kt
/*
 * Copyright (c) 2020. BoostTag E.I.R.L. Romell D.Z.
 * All rights reserved
 * porfile.romellfudi.com
 */

package com.romellfudi.ussd.sample

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.romellfudi.permission.PermissionService
import com.romellfudi.ussd.R
import com.romellfudi.ussd.di.component.DaggerMainFragmentComponent
import com.romellfudi.ussd.di.module.MainFragmentModule
import com.romellfudi.ussdlibrary.OverlayShowingService
import com.romellfudi.ussdlibrary.SplashLoadingService
import com.romellfudi.ussdlibrary.USSDApi
import com.romellfudi.ussdlibrary.USSDController
import kotlinx.android.synthetic.main.content_op1.*
import java.util.*
import javax.inject.Inject

/**
 * Use Case for Test Windows
 *
 * @author Romell Domínguez
 * @version 1.1.b 27/09/2018
 * @since 1.0.a
 */

class MainFragment : Fragment() {

    private lateinit var map: HashMap<String, HashSet<String>>

    @Inject
    lateinit var ussdApi: USSDApi

    @Inject
    lateinit var permissionService: PermissionService

    private val callback = object : PermissionService.Callback() {
        override fun onResponse(refusePermissions: ArrayList<String>?) {
            Toast.makeText(context,
                    getString(R.string.refuse_permissions),
                    Toast.LENGTH_SHORT).show()
            activity!!.finish()
        }

    }

    @SuppressLint("MissingPermission")
    override fun onCreate(savedInstanceState: Bundle?) {
//        (activity!!.application as App).appComponent.uiComponent().create().inject(this)
        DaggerMainFragmentComponent.builder()
//                .appComponent((activity!!.application as App).initializeComponent())//appComponent.uiComponent().create())
                .mainFragmentModule(MainFragmentModule(activity!!))
                .build().inject(this)

        super.onCreate(savedInstanceState)

        map = HashMap()
        map["KEY_LOGIN"] = HashSet(listOf("espere", "waiting", "loading", "esperando"))
        map["KEY_ERROR"] = HashSet(listOf("problema", "problem", "error", "null"))

        permissionService.request(callback)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.content_op1, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setHasOptionsMenu(false)

        btn1.setOnClickListener {
            val phoneNumber = phone!!.text.toString().trim { it <= ' ' }
            ussdApi = USSDController.getInstance(activity!!)
            result.text = ""
            ussdApi.callUSSDInvoke(phoneNumber, map!!,
                    object : USSDController.CallbackInvoke {
                        override fun responseInvoke(message: String) {
                            Log.d("APP", message)
                            result.append("\n-\n$message")
                            // first option list - select option 1
                            ussdApi.send("1") {
                                Log.d("APP", it)
                                result.append("\n-\n$it")
                                // second option list - select option 1
                                ussdApi.send("1") {
                                    Log.d("APP", it)
                                    result.append("\n-\n$it")
                                }
                            }
//                            ussdApi.cancel()
                        }

                        override fun over(message: String) {
                            Log.d("APP", message)
                            result.append("\n-\n$message")
                        }
                    })
        }

        btn2.setOnClickListener {
            if (USSDController.verifyOverLay(activity!!)) {
                val svc = Intent(activity, OverlayShowingService::class.java)
                svc.putExtra(OverlayShowingService.EXTRA, "PROCESANDO")
                activity!!.startService(svc)
                Log.d("APP", "START OVERLAY DIALOG")
                val phoneNumber = phone!!.text.toString().trim { it <= ' ' }
                ussdApi = USSDController.getInstance(activity!!)
                result.text = ""
                ussdApi.callUSSDOverlayInvoke(phoneNumber, map!!,
                        object : USSDController.CallbackInvoke {
                            override fun responseInvoke(message: String) {
                                Log.d("APP", message)
                                result.append("\n-\n$message")
                                // first option list - select option 1
                                ussdApi.send("1") {
                                    Log.d("APP", it)
                                    result.append("\n-\n$it")
                                    // second option list - select option 1
                                    ussdApi.send("1") {
                                        Log.d("APP", it)
                                    }
                                }
//                                ussdApi.cancel()
                            }

                            override fun over(message: String) {
                                Log.d("APP", message)
                                result.append("\n-\n$message")
                                activity!!.stopService(svc)
                                Log.d("APP", "STOP OVERLAY DIALOG")
                            }
                        })
            }
        }

        btn4.setOnClickListener {
            if (USSDController.verifyOverLay(activity!!)) {
                val svc = Intent(activity, SplashLoadingService::class.java)
                activity!!.startService(svc)
                Log.d("APP", "START SPLASH DIALOG")
                val phoneNumber = phone!!.text.toString().trim { it <= ' ' }
                result.text = ""
                ussdApi.callUSSDOverlayInvoke(phoneNumber, map,
                        object : USSDController.CallbackInvoke {
                            override fun responseInvoke(message: String) {
                                Log.d("APP", message)
                                result.append("\n-\n$message")
                                // first option list - select option 1
                                ussdApi.send("1") {
                                    Log.d("APP", it)
                                    result.append("\n-\n$message")
                                    // second option list - select option 1
                                    ussdApi.send("1") {
                                        Log.d("APP", it)
                                        result.append("\n-\n$it")
                                        activity!!.stopService(svc)
                                        Log.d("APP", "STOP SPLASH DIALOG")
                                        Log.d("APP", "successful")
                                    }
                                }
//                                ussdApi.cancel()
                            }

                            override fun over(message: String) {
                                Log.d("APP", message)
                                result!!.append("\n-\n$message")
                                activity!!.stopService(svc)
                                Log.d("APP", "STOP OVERLAY DIALOG")
                            }
                        })
            }
        }

        btn3.setOnClickListener(fun(it: View) {
            USSDController.verifyAccesibilityAccess(activity!!)
        })

    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>,
                                            grantResults: IntArray) =
            callback.handler(permissions, grantResults)

}